### PR TITLE
Uniqueness validations

### PR DESCRIPTION
**Pull request by coderdan**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -274,6 +274,28 @@ User.reindex
...
+Standard ActiveRecord uniqueness validations won't work when data is encrypted.
+While it *can* work in some cases when using deterministic encryption, we generally recommend against this approach
+(deterministic encryption is vulnerable to inference and chosen-plaintext attacks).
+
+Instead, you can include `ActiveStash::Validations` and your uniqueness validation will now work via a query
+to the CipherStash index on the validated field.
+If there is no index on the field then validations will fail.
+
+Also, note that as of now the `ActiveStash::Validations` module only supports a very simple uniqueness validation
+and doesn't yet support conditions or any option other than `:message`.
```
Comment (mpalmer) [inline README.md:288]: Ah, that's going to be insufficient for [the initial use-case I'm after this feature for](https://github.com/heartcombo/devise/blob/v4.7.3/lib/devise/models/validatable.rb#L36).

File: lib/active_stash/validations.rb [hunk 2]
```diff
@@ -0,0 +1,11 @@
+module ActiveStash
+  module Validations
+     class UniquenessValidator < ActiveModel::EachValidator
+    def validate_each(record, attribute, value)
+      if record.class.query(attribute => value).exists?
+        record.errors.add(attribute, options[:message] || "already exists")
+      end
+    end
+  end
+  end
```
Comment (mpalmer) [inline lib/active_stash/validations.rb:10]: Looks like some wonky indenting going on there...

```suggestion
    class UniquenessValidator < ActiveModel::EachValidator
      def validate_each(record, attribute, value)
        if record.class.query(attribute => value).exists?
          record.errors.add(attribute, options[:message] || "already exists")
        end
      end
    end
  end
```

**Review by mpalmer — DISMISSED** [review]
Nothing terribad in here that I noticed, so approving, but it won't be sufficient to support the devise-instantiated uniqueness validations I'm wrestling with at the moment.

**Review by mpalmer — COMMENTED** [review]
Your code looks OK, but when I try to use it with a Devise-using model it isn't being picked up, and the SQL-based validation still runs (and explodes, ofc).  I'll have to spend some time figuring out what strange magicks Devise is up to, and get back to you.

File: README.md [hunk 3]
```diff
@@ -274,6 +274,31 @@ User.reindex
...
 
+## Uniqueness Validations
+
+Standard ActiveRecord uniqueness validations won't work when data is encrypted.
+While it *can* work in some cases when using deterministic encryption, we generally recommend against this approach
+(deterministic encryption is vulnerable to inference and chosen-plaintext attacks).
+
+Instead, you can include `ActiveStash::Validations` and your uniqueness validation will now work via a query
+to the CipherStash index on the validated field.
+If there is no index on the field then validations will fail.
```
[the remark this answers is not in the document]
Comment (coderdan) [reply README.md:285]: Not sure. Will look into that.

File: lib/active_stash/relation.rb [hunk 4]
```diff
@@ -50,6 +50,11 @@ def order(*args)
       self
     end
 
+    def exists?
+      self.load
```
[the remark this answers is not in the document]
Comment (coderdan) [reply lib/active_stash/relation.rb:54]: Supposedly, yes but it wasn't working without this line. I'll double check but something weird with lazy loading.

File: spec/support/user.rb [hunk 5]
```diff
@@ -1,9 +1,19 @@
...
   include ActiveStash::Search
+  include ActiveStash::Validations
   self.table_name = "users"
   self.collection_name = "activestash_test_#{ENV["ACTIVE_STASH_TEST_COLLECTION_PREFIX"] || ""}_users"
 
-  stash_index :first_name, :dob, :created_at
+  # Used for testing
+  attr_accessor :skip_validations
+
+  validates :email, uniqueness: true, if: Proc.new { |user| user.perform_validations? }
```
Comment (mpalmer) [inline spec/support/user.rb:10]:
```suggestion
  validates_uniqueness_of :email, if: Proc.new { |user| user.perform_validations? }
```

Finally figured out WTF is going on in Devise.  Turns out that `validates_uniqueness_of` is what Devise is calling, and that doesn't end up getting turned into a `validates uniqueness: true` call, but instead does something totally different.  NFI *what*, yet, but this at least gives us a failing test case to work from.

**Review by mpalmer — CHANGES_REQUESTED** [review]

File: README.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (fimac) [reply README.md:292]: I'll add a test and check this out.

File: lib/active_stash/validations.rb (hunk 2 above)
Comment (fimac) [reply lib/active_stash/validations.rb]: Looks like this has been addressed.

File: README.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (fimac) [reply README.md:292]: @mpalmer @coderdan I tested this out, and ActiveStash::Validations overrides all ActiveRecord uniqueness validations.

In this example, first_name doesn't have a stash_index applied, and has a uniqueness validation.


https://github.com/cipherstash/activestash/compare/uniqueness-validations...test-overrides-on-non-encrypted-indexes?expand=1

https://github.com/cipherstash/activestash/compare/uniqueness-validations...test-overrides-on-non-encrypted-indexes?expand=1

The error returned is `expected no Exception, got #<RuntimeError: No indexes available for 'first_name'> with backtrace:`

Screen Shot 2022-08-12 at 11 52 56 am

[the remark this answers is not in the document]
Comment (fimac) [reply README.md:292]: I'll try something like the below pseudo code and see if I can get it working. 

```ruby
module ClassMethods
      def validates_uniqueness_of(*attr_names)
        # if stash index available for attr_name
        validates_with ActiveStash::Validations::UniquenessValidator, _merge_attributes(attr_names)
        # else
        validates_with UniquenessValidator, _merge_attributes(attr_names)
        # end
      end
    end
```

[the remark this answers is not in the document]
Comment (fimac) [reply README.md:323]: This issue has been fixed with this change
https://github.com/cipherstash/activestash/pull/46/commits/edbcba895307a73a09a206b4f62bad28aaf4bd7c

**Review by mpalmer — APPROVED** [review]

**Pull request merged**